### ISSUE TITLE
rule no-undefined-types does not work

### ISSUE THREAD

**Issue by DarkPark** [gajus/eslint-plugin-jsdoc#109]
eslint version: 5.9.0
eslint-plugin-jsdoc version: 3.9.1
Config:
```
'jsdoc/check-param-names': 1,
'jsdoc/check-tag-names': 1,
'jsdoc/check-types': 1,
'jsdoc/newline-after-description': 1,
'jsdoc/no-undefined-types': 1,
'jsdoc/require-description': 0,
'jsdoc/require-description-complete-sentence': 0,
'jsdoc/require-example': 0,
'jsdoc/require-hyphen-before-param-description': 1,
'jsdoc/require-param': 1,
'jsdoc/require-param-description': 1,
'jsdoc/require-param-name': 1,
'jsdoc/require-param-type': 1,
'jsdoc/require-returns-description': 1,
'jsdoc/require-returns-type': 1,
'jsdoc/valid-types': 1
```

Your own example:

```js
const MyType = require('my-library').MyType;

/**
 * @param {MyType} foo - Bar.
 */
function quux(foo) {

}
```
gives the error `warning  The type 'MyType' is undefined  jsdoc/no-undefined-types`

**Comment by bary12** [comment]
How come the test passes but the actual behavior is different? Is this a problem with the test suite?

**Comment by brettz9** [comment]
I'm no longer seeing this issue. @DarkPark , can you confirm?

**Comment by brettz9** [comment]
(This test is now a part of the suite, so I assume it has since been fixed.)

**Comment by DarkPark** [comment]
@brettz9 unfortunately the problem remains.

```bash
$ eslint/bin/eslint.js -v
v5.16.0
```

Run check with the [config.js](https://pastebin.com/xDMDmTpZ):

```bash
$ eslint/bin/eslint.js -c config.js test.js
v5.16.0
```

Result:

```
test.js
  3:7   error    'MyType' is assigned a value but never used  no-unused-vars
  6:0   warning  The type 'MyType' is undefined               jsdoc/no-undefined-types
  8:10  error    'quux' is defined but never used             no-unused-vars
  8:17  error    'foo' is defined but never used              no-unused-vars
  8:23  error    Unexpected empty function 'quux'             no-empty-function

✖ 5 problems (4 errors, 1 warning)
```

There are some other unrelated expected issues here, please ignore them.

**Comment by brettz9** [comment]
Did you check against `master`? The latest commit was made 6 days ago.

**Comment by DarkPark** [comment]
no
this check was done with the latest npm version

**Comment by brettz9** [comment]
Ok, well, you can try changing to the following (and running `npm install`) to confirm:

```json
"devDependencies": {
    "eslint-plugin-jsdoc": "https://github.com/gajus/eslint-plugin-jsdoc"
}
```

**Comment by DarkPark** [comment]
strange but your advice doesn't work at all

```
Oops! Something went wrong! :(

ESLint: 5.16.0.
ESLint couldn't find the plugin "eslint-plugin-jsdoc". This can happen for a couple different reasons:
...
```

There are only 3 files in node_modules/eslint-plugin-jsdoc:

```
LICENSE
package.json
README.md
```

**Comment by DarkPark** [comment]
anyway I've used npm package "eslint-plugin-jsdoc": "^4.8.4"
it was released 3 days ago

**Comment by brettz9** [comment]
Oh, my apologies! 4.8.4 should be ok and the Github address won't work as the `dist` files aren't on Github.

**Comment by brettz9** [comment]
And you ran `npm i` against 4.8.4?

**Comment by DarkPark** [comment]
yes

```
    "dependencies": {
        "eslint": "^5.16.0",
        "eslint-plugin-jsdoc": "^4.8.4"
    }
```

**Comment by brettz9** [comment]
I set up a local repository with this `.eslintrc.js` config:

```js
module.exports = {
    parserOptions: {
        "ecmaVersion": 6
    },
    plugins: ['jsdoc'],
    rules: {
        'jsdoc/check-param-names': 1,
        'jsdoc/check-tag-names': 1,
        'jsdoc/check-types': 1,
        'jsdoc/newline-after-description': 1,
        'jsdoc/no-undefined-types': 1,
        'jsdoc/require-description': 0,
        'jsdoc/require-description-complete-sentence': 0,
        'jsdoc/require-example': 0,
        'jsdoc/require-hyphen-before-param-description': 1,
        'jsdoc/require-param': 1,
        'jsdoc/require-param-description': 1,
        'jsdoc/require-param-name': 1,
        'jsdoc/require-param-type': 1,
        'jsdoc/require-returns-description': 1,
        'jsdoc/require-returns-type': 1,
        'jsdoc/valid-types': 1
    }
}
```

and this test file, `test.js`:

```js
const MyType = require('my-library').MyType;

/**
 * @param {MyType} foo - Bar.
 */
function quux(foo) {

}
```

and this `package.json` with all the init defaults besides `dependencies`:

```json
{
  "name": "eslint-plugin-jsdoc-test-109",
  "version": "1.0.0",
  "description": "",
  "main": ".eslintrc.js",
  "scripts": {
    "eslint": "eslint .",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "eslint": "^5.16.0",
    "eslint-plugin-jsdoc": "^4.8.4"
  }
}
```

and running `./node_modules/.bin/eslint test.js` from the command line gives no errors.

**Comment by brettz9** [comment]
Just making sure you actually ran `npm install` again after changing it in `dependencies`, as one can sometimes forget to do this...

**Comment by DarkPark** [comment]
@brettz9 thanks
your example helped to find a culprit
in my config i use environment setup:

```
env: {
    commonjs: true,
    es6: true,
    node: true
}
```

commenting these two lines:

```
commonjs: true,
node: true
```

resolves the issue though I don't understand how it works

**Comment by brettz9** [comment]
Ok, and I see `jsdoc/no-undefined-types': 1` as the sole rule is enough to show the problem. And it is apparently because in Node, `const` doesn't set variables as globals, which is what the rule is checking for. I'll see about taking a look.

**Comment by DarkPark** [comment]
thank you, as I really need to use `node` env

**Comment by brettz9** [comment]
I've submitted PR #222 as a fix; it is now passing in Node/commonjs in my testing added to the PR, so we'll see about it possibly getting merged.

**Comment by DarkPark** [comment]
thanks

**Comment by DarkPark** [comment]
it works
thank you

**Comment by brettz9** [comment]
Excellent!
